issuance/crl.go

```diff
@@ -0,0 +1,165 @@
+
+	// TODO(#7296): Remove this and instead compute it from Issuer.CRLBaseURL
+	DeprecatedIDPBaseURL string
+}
+
+func (i *Issuer) IssueCRL(prof *CRLProfile, req *CRLRequest) ([]byte, error) {
+	backdatedBy := i.clk.Now().Sub(req.ThisUpdate)
+	if backdatedBy > prof.maxBackdate {
+		return nil, fmt.Errorf("ThisUpdate is too far in the past (%s>%s)", backdatedBy, prof.maxBackdate)
+	}
+	if backdatedBy < 0 {
+		return nil, fmt.Errorf("ThisUpdate is in the future (%s>%s)", req.ThisUpdate, i.clk.Now())
+	}
+
+	template := &x509.RevocationList{
+		RevokedCertificateEntries: req.Entries,
+		Number:                    req.Number,
+		ThisUpdate:                req.ThisUpdate,
+		NextUpdate:                req.ThisUpdate.Add(-time.Second).Add(prof.validityInterval),
+	}
+
+	if i.crlURLBase == "" && req.DeprecatedIDPBaseURL == "" {
+		return nil, fmt.Errorf("CRL must contain an issuingDistributionPoint")
+	}
+
+	var idps []string
+	if i.crlURLBase != "" {
+		idps = append(idps, fmt.Sprintf("%s/%d.crl", i.crlURLBase, req.Shard))
+	}
+	if req.DeprecatedIDPBaseURL != "" {
+		// TODO(#7296): Remove this fallback once CCADB and all non-expired certs
+		// contain the new-style CRLDP URL instead.
+		idps = append(idps, fmt.Sprintf("%s/%d/%d.crl", req.DeprecatedIDPBaseURL, i.NameID(), req.Shard))
+	}
+	idp, err := makeIDPExt(idps)
+	if err != nil {
+		return nil, fmt.Errorf("creating IDP extension: %w", err)
+	}
+	template.ExtraExtensions = append(template.ExtraExtensions, idp)
+
+	err = i.Linter.CheckCRL(template, prof.lints)
+	if err != nil {
+		return nil, err
+	}
+
+	crlBytes, err := x509.CreateRevocationList(
+		rand.Reader,
+		template,
+		i.Cert.Certificate,
+		i.Signer,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return crlBytes, nil
+}
+
+// distributionPointName represents the ASN.1 DistributionPointName CHOICE as
+// defined in RFC 5280 Section 4.2.1.13. We only use one of the fields, so the
+// others are omitted.
+type distributionPointName struct {
+	// Technically, FullName is of type GeneralNames, which is of type SEQUENCE OF
+	// GeneralName. But GeneralName itself is of type CHOICE, and the ans1.Marhsal
+	// function doesn't support marshalling structs to CHOICEs, so we have to use
+	// asn1.RawValue and encode the GeneralName ourselves.
+	FullName []asn1.RawValue `asn1:"optional,tag:0"`
+}
+
+// issuingDistributionPoint represents the ASN.1 IssuingDistributionPoint
+// SEQUENCE as defined in RFC 5280 Section 5.2.5. We only use two of the fields,
+// so the others are omitted.
+type issuingDistributionPoint struct {
+	DistributionPoint     distributionPointName `asn1:"optional,tag:0"`
+	OnlyContainsUserCerts bool                  `asn1:"optional,tag:1"`
+}
+
+// makeIDPExt returns a critical IssuingDistributionPoint extension containing
+// the given URLs and with the OnlyContainsUserCerts boolean set to true.
+func makeIDPExt(urls []string) (pkix.Extension, error) {
+	gns := make([]asn1.RawValue, len(urls))
+	for i, url := range urls {
```

Comment:
Stylistically, I prefer to start with an empty slice and append to it, presuming this isn't performance-critical or doing something cute with concurrency.

Second choice is to do `make([]asn1.RawValue, 0, len(urls))` and append to it. That reduce the chance that later refactorings could create a path where some of the preallocated zero-valued URLs were never assigned to.

Reply:
This is a good point, I actually had exactly that bug halfway through writing this PR. Changed to use append on top of an empty slice.